### PR TITLE
Add workaround for Chrome/Edge css import escaping bug

### PR DESCRIPTION
**Pull request by Juice10**
Edge and Chrome currently have a bug when escaping import statements which can break recording (and playback): https://bugs.chromium.org/p/chromium/issues/detail?id=1472259

This PR adds a workaround for this, and upgrades rrweb to typescript 4.9.5

### CODE REVIEW

File: packages/rrweb-snapshot/test/__snapshots__/integration.test.ts.snap [hunk 1]
```diff
@@ -490,7 +490,7 @@ exports[`integration tests [html file]: with-style-sheet-with-import.html 1`] =
...
+  <style>@import url(\\"//fonts.googleapis.com/css2?family=Open+Sans:wght@400;600;700&amp;family=Roboto:wght@100;300;400;500;700&amp;display=swap\\\\"\\");body { margin: 0px; background: url(\\"http://localhost:3032/a.jpg\\"); border-image: url(\\"data:image/svg+xml;utf8,&lt;svg xmlns=\\\\\\"http://www.w3.org/2000/svg\\\\\\" x=\\\\\\"0px\\\\\\" y=\\\\\\"0px\\\\\\" viewBox=\\\\\\"0 0 256 256\\\\\\"&gt;&lt;g&gt;&lt;g&gt;&lt;polygon points=\\\\\\"79.093,0 48.907,30.187 146.72,128 48.907,225.813 79.093,256 207.093,128\\\\\\"/&gt;&lt;/g&gt;&lt;/g&gt;&lt;/svg&gt;\\") 100% / 1 / 0 stretch; }p { color: red; background: url(\\"http://localhost:3032/css/b.jpg\\"); }body &gt; p { color: yellow; }</style>
```
Comment (Juice10) [inline packages/rrweb-snapshot/test/__snapshots__/integration.test.ts.snap:493]:
```suggestion
  <style>@import url(\\"//fonts.googleapis.com/css2?family=Open+Sans:wght@400;600;700&amp;family=Roboto:wght@100;300;400;500;700&amp;display=swap\\\\"\\");body { margin: 0px; background: url(\\"http://localhost:3030/a.jpg\\"); border-image: url(\\"data:image/svg+xml;utf8,&lt;svg xmlns=\\\\\\"http://www.w3.org/2000/svg\\\\\\" x=\\\\\\"0px\\\\\\" y=\\\\\\"0px\\\\\\" viewBox=\\\\\\"0 0 256 256\\\\\\"&gt;&lt;g&gt;&lt;g&gt;&lt;polygon points=\\\\\\"79.093,0 48.907,30.187 146.72,128 48.907,225.813 79.093,256 207.093,128\\\\\\"/&gt;&lt;/g&gt;&lt;/g&gt;&lt;/svg&gt;\\") 100% / 1 / 0 stretch; }p { color: red; background: url(\\"http://localhost:3030/css/b.jpg\\"); }body &gt; p { color: yellow; }</style>
```

File: packages/rrweb/src/record/observers/canvas/canvas.ts [hunk 2]
```diff
@@ -19,19 +19,19 @@ export default function initCanvasContextObserver(
       'getContext',
       function (
         original: (
-          this: ICanvas,
+          this: ICanvas | HTMLCanvasElement,
```
Comment (eoghanmurray) [inline packages/rrweb/src/record/observers/canvas/canvas.ts:22]: Is this Canvas stuff in here by mistake?

Comment (Juice10) [reply packages/rrweb/src/record/observers/canvas/canvas.ts:22]: No its due to typescript update, its gotten stricter

Comment (Juice10) [reply packages/rrweb/src/record/observers/canvas/canvas.ts:22]: This diff contains all the changes related to typescript 4.9.5: https://github.com/rrweb-io/rrweb/pull/1287/commits/bf6d3f311d8506f94608793f56b9abb248cd7979

I've had to upgrade typescript to 4.9.5 as the lib.dom.d.ts contains a number of things this PR uses

File: packages/rrweb-snapshot/src/utils.ts [hunk 3]
```diff
@@ -71,7 +111,14 @@ export function getCssRuleString(rule: CSSRule): string {
   let cssStringified = rule.cssText;
   if (isCSSImportRule(rule)) {
     try {
-      cssStringified = getCssRulesString(rule.styleSheet) || cssStringified;
+      cssStringified =
+        getCssRulesString(rule.styleSheet) ||
+        // browser import didn't work due to CORS
+        // so we try to construct the import statement ourselves
+        // and check for browser compatibility issues
+        fixBrowserCompatibilityIssuesInCSSImports(rule) ||
```
[the remark this answers is not in the document]
Comment (eoghanmurray) [reply packages/rrweb-snapshot/src/utils.ts:119]: I'm failing to follow which `null` it is detecting out of the two possible code paths that return null in `getCssRulesString`.. I think the `fixBrowserCompatibilityIssuesInCSS` should be moved closer to the bit that triggers it

**Review by eoghanmurray — APPROVED** [review]

**Pull request merged**